nixos/test-driver: provide a global timeout

RaitoBezarius opened the pull request:
## Description of changes

Since the debut of the test-driver, we didn't obtain a race timer with the test execution to ensure that tests doesn't run beyond a certain amount of time.

This is particularly important when you are running into hanging tests which cannot be detected by current facilities (requires more pvpanic wiring up, QMP API stuff, etc.).

Two easy examples:

- Some QEMU tests may get stuck in some situation and run for more than 24 hours → we default to 24 hours max.
- Some QEMU tests may panic in the wrong place, e.g. UEFI firmware or worse → end users can set a "reasonable" amount of time

And then, we should let the retry logic retest them until they succeed and adjust their global timeouts.

Of course, this does not help with the fact that the timeout may need to be a function of the actual busyness of the machine running the tests. This is only one step towards increased reliability.

Self tests are offered via `testBuildFailure`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nixos/test-driver: provide a global timeout

Since the debut of the test-driver, we didn't obtain
a race timer with the test execution to ensure that tests doesn't run beyond
a certain amount of time.

This is particularly important when you are running into hanging tests
which cannot be detected by current facilities (requires more pvpanic wiring up, QMP
API stuff, etc.).

Two easy examples:

- Some QEMU tests may get stuck in some situation and run for more than 24 hours → we default to 1 hour max.
- Some QEMU tests may panic in the wrong place, e.g. UEFI firmware or worse → end users can set a "reasonable" amount of time

And then, we should let the retry logic retest them until they succeed and adjust
their global timeouts.

Of course, this does not help with the fact that the timeout may need to be
a function of the actual busyness of the machine running the tests.
This is only one step towards increased reliability.

roberth commented:
:+1: this is more robust than `meta.timeout`, e.g. when tests are accumulated with `symlinkJoin` or some reporting derivation.
Would love to use this in Nix itself to report which tests have run in which test environments. That's currently far too opaque.

> TODO: self tests?

You could write a test with a short timeout and a long sleep somewhere, then transform it with [`testers.testBuildFailure`](https://nixos.org/manual/nixpkgs/unstable/#tester-testBuildFailure), which produces a log in its `$out`.
You can then grep the log from a `runCommand` to make sure it failed in the right way (ie not due to an unrelated error).

tfc commented:
General timeouts are useful and generally an improvement for CI throughput.

I have two thoughts about this change:

For how many people is a test without a timeout useful at all? In interactive mode (which could run for days for playing with), the timer is not started, which makes sense.
As much as i like "maybe" values, we could get rid of this  complexity by simply always having a timeout. And people who want no timeout can still simulate it by setting it to a week or so (and interactive users don't need to touch it at all).

The default value of 24 hours is already very relaxed. What is our knowledge about the longest legitimately running test? It might make sense to already use this CI-throughput-improving functionality by setting a useful halfway strict global timeout (as it still can be set to longer times for tests which really need it).

RaitoBezarius commented:
> General timeouts are useful and generally an improvement for CI throughput.
> 
> I have two thoughts about this change:
> 
> For how many people is a test without a timeout useful at all? In interactive mode (which could run for days for playing with), the timer is not started, which makes sense.
> As much as i like "maybe" values, we could get rid of this  complexity by simply always having a timeout. And people who want no timeout can still simulate it by setting it to a week or so (and interactive users don't need to touch it at all).

Sounds reasonable!

> 
> The default value of 24 hours is already very relaxed. What is our knowledge about the longest legitimately running test? It might make sense to already use this CI-throughput-improving functionality by setting a useful halfway strict global timeout (as it still can be set to longer times for tests which really need it).

Discussed it on infra and 1 hour seems better.

nikstur commented on nixos/lib/test-driver/test_driver/__init__.py:
```diff
@@ -76,6 +76,14 @@ def main() -> None:
+        metavar="GLOBAL_TIMEOUT",
+        action=EnvDefault,
+        envvar="globalTimeout",
+        help="Global timeout in seconds to arm for the whole test",
```
```suggestion
        help="Timeout in seconds for the whole test",
```

nikstur commented on nixos/lib/testing/driver.nix:
```diff
@@ -123,6 +124,18 @@ in
+        A global timeout for the complete test, expressed in seconds.
+        Beyond that timeout, every resource will be killed and released and the test will fail.
+
+        By default, we use a 24 hours timeout.
```
24 sounds very long. What about 6 hours? Do we have any tests that run longer? If they do it's probably better if they manually override the timeout anyway.

nikstur reviewed (commented):
Very very cool

blitz reviewed (approved):
I love it! :heart:

RaitoBezarius commented:
For self testing, I am reviving #157161 and plan to use it as a base for the self test here.

RaitoBezarius committed:
nixos/lib/testing/run: expose `rawTestDerivation`

For `testBuildFailure` and similar functions, we need a full blown derivation and not a lazy one.
This is an internal option for test framework developers.

RaitoBezarius committed:
nixos/tests/nixos-test-driver: test timeout failures

We test that the test framework timeouts are working as expected.

RaitoBezarius committed:
nixos/lib/test-driver: add driver-timeout as a passthru test

From now on, we will aim to ensure that the test driver
gets tested by OfBorg using all our available tests.

This commit adds the driver timeout test to the driver.

RaitoBezarius commented:
@roberth Done for everything in this scope. I would suggest to get this first probably. :)

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -161,6 +172,12 @@ def join_all(self) -> None:
+            self.race_timer.cancel()
+
+    def terminate_test(self) -> None:
+        with rootlog.nested("timeout reached; test termination..."):
```
```suggestion
        with rootlog.nested("timeout reached; test terminating..."):
```

roberth commented on nixos/tests/nixos-test-driver/timeout.nix:
```diff
@@ -0,0 +1,18 @@
+
+  nodes = {
+    machine = ({ pkgs, ... }: {
+      environment.systemPackages = [
+        pkgs.busybox
+      ];
```
```suggestion
```

Don't we have `sleep` from `coreutils`?

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -144,6 +151,10 @@ def test_script(self) -> None:
     def run_tests(self) -> None:
         """Run the test script (for non-interactive test runs)"""
+        rootlog.info(
+            f"Test will timeout and terminate in {self.global_timeout} seconds"
```
```suggestion
            f"Test will time out and terminate in {self.global_timeout} seconds"
```

verb

roberth commented on nixos/lib/test-driver/test_driver/driver.py:
```diff
@@ -161,6 +172,12 @@ def join_all(self) -> None:
+    def terminate_test(self) -> None:
+        with rootlog.nested("timeout reached; test termination..."):
+            for machine in self.machines:
+                machine.release()
```
This doesn't seem to handle non-termination in the `testScript`.
Also what happens if the `testScript` has a `try`/`except`? The option doc claims that the script will fail, but this does not seem to be guaranteed.

RaitoBezarius replied:
What do you mean exactly by non-termination? The test passes? Termination is done somewhere else (by releasing all the machines too with a different message).

Good point I guess about the `try/except`, I wonder if there's any form of non-catchable exception inside Python but everyone would break down in general.

In any case, I guess we could ensure termination by exiting with exit code 2 (to make it clear it's a timeout-driven exit).

RaitoBezarius replied:
Here's a proposal, I will kill the main thread in the termination sequence.

RaitoBezarius replied:
(more precisely: I will kill the process and let the kernel do the right thing.)

roberth replied:
> kill the process

sgtm.

Alternatively, python may have a direct exit "syscall" that doesn't throw an exception but just does the thing immediately.

roberth replied:
This seems to be it: [os._exit](https://docs.python.org/3/library/os.html#os._exit)

RaitoBezarius replied:
Oh right, I guess using `kill` is less obscure than `_exit`, if that's OK with you.

roberth replied:
Not sure if that's so great though. `sys.exit` might throw an exception and therefore not be airtight, but you don't get any cleanup either. Script author won't be able to clean up out-of-process resources, such as child processes.
Maybe it should just be `sys.exit` anyway, although I don't know how that exception would propagate to the main thread?

roberth replied:
Usually we have Nix to fall back on as a kind of process manager, but that's not the case with interactive use (`nix run .#foo.driverInteractive`).

RaitoBezarius replied:
For non-interactive usecase, this is exactly what we want, I would say.
For interactive usecase, well… it was ruled out of scope from the get-go and I would argue that timeout should be disabled in the interactive usecase.

Using `sys.exit` is impossible because there's no way we can propagate the `threading.Timer` exception back to the main thread.

We would need our own way to control error handling and offer a chance to terminate gracefully, I would expect this would require some serious back'n'forth in the code to achieve this, including running the test script on its own thread and having the main thread as a manager.

AFAIK, out of process resources should still be killed in interactive because they are spawned off from a unique hierarchy (the interactive runner).

So… I would argue this is as good as we can go without investing much more time, which I am afraid I won't be able to commit to this in particular.

roberth replied:
> there's no way we can propagate

Makes sense actually. Usually only languages that exercise tight control over their runtime allow this, examples being Java and Haskell.
(Also pthreads I guess, but unsafely)

Sounds good enough to me then.

RaitoBezarius commented on nixos/lib/testing/run.nix:
```diff
@@ -29,25 +38,26 @@ in
-        requiredSystemFeatures = [ "kvm" "nixos-test" ];
+      requiredSystemFeatures = [ "kvm" "nixos-test" ];
 
-        buildCommand = ''
```
Ugh, I should maybe get rid of those unrelated changes.

RaitoBezarius replied:
Ah they are needed…
